pkg/metrics/metrics.go

```diff
@@ -0,0 +1,255 @@
+
+	// textFileCollector uses a hash of pod UID and volume name as an identifier.
+	c := NewTextFileCollector("", podUID, volumeName, nil)
+	if ok := mm.registry.Unregister(c); !ok {
+		klog.Infof("Unregister metrics collector for targetPath %q is not needed since the collector is not registered", targetPath)
+	}
+}
+
+type textFileCollector struct {
+	path        string
+	constLabels map[string]string
+	podUID      string
+	volumeName  string
+}
+
+// NewTextFileCollector returns a new Collector exposing metrics read from the give path.
+func NewTextFileCollector(path, podUID, volumeName string, labels map[string]string) prometheus.Collector {
+	c := &textFileCollector{
+		path:        path,
+		constLabels: labels,
+		podUID:      podUID,
+		volumeName:  volumeName,
+	}
+
+	return c
+}
+
+// Describe emits the description of metrics.
+// Prometheus Registry relies on this func to identify collectors.
+func (c *textFileCollector) Describe(ch chan<- *prometheus.Desc) {
+	// Collector id is a hash of the values of the ConstLabels and fqName.
+	ch <- prometheus.NewDesc("gke_gcsfuse_csi_metric", "GKE GCSFuse CSI metric.", nil, map[string]string{"pod_uid": c.podUID, "volume_name": c.volumeName})
+}
+
+// Collect emits metrics.
+func (c *textFileCollector) Collect(ch chan<- prometheus.Metric) {
+	families, err := ProcessMetricsFile(c.path)
+	if err != nil {
+		klog.Errorf("failed to process metrics from metrics file: %v", err)
+
+		return
+	}
+
+	for _, mf := range families {
+		c.emitMetricFamily(mf, ch)
+	}
+}
+
+// ProcessMetricsFile processes a metrics file that follows Prometheus text format: https://prometheus.io/docs/instrumenting/exposition_formats/,
+// returning its MetricFamily.
+func ProcessMetricsFile(path string) (map[string]*dto.MetricFamily, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open metrics file %q: %w", path, err)
+	}
+	defer f.Close()
+
+	var parser expfmt.TextParser
+	metricFamilies, err := parser.TextToMetricFamilies(f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse metrics file %q: %w", path, err)
+	}
+
+	return metricFamilies, nil
+}
+
+// emitMetricFamily iterates MetricFamily, converts metricFamily.Metric to prometheus.Metric, and emits the metric via the given chan.
+func (c *textFileCollector) emitMetricFamily(metricFamily *dto.MetricFamily, ch chan<- prometheus.Metric) {
+	var valType prometheus.ValueType
+	var val float64
+
+	for _, metric := range metricFamily.GetMetric() {
+		var LabelNames []string
+		var LabelValues []string
+		for _, label := range metric.GetLabel() {
+			LabelNames = append(LabelNames, label.GetName())
+			LabelValues = append(LabelValues, label.GetValue())
+		}
+
+		for n, v := range c.constLabels {
+			LabelNames = append(LabelNames, n)
+			LabelValues = append(LabelValues, v)
+		}
+
+		emitNewConstMetric := func() {
+			ch <- prometheus.MustNewConstMetric(
+				prometheus.NewDesc(
+					metricFamily.GetName(),
```

Comment:
@hime Here is the place you can find the metric name, and you can modify the metric label `fs_error` in the metric `fs_ops_error_count`.

Reply:
Closing as `fs_error` label was completely dropped from gcsfuse.